TreeGrid keyboard navigation

elmot opened the pull request:
Fixes #8758

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/vaadin/framework/8906)
<!-- Reviewable:end -->

Commit:
TreeGrid keyboard navigation

Fixes #8758

hesara commented:
Reviewed 1 of 5 files at r1.
Review status: 1 of 5 files reviewed at latest revision, 7 unresolved discussions.

---

*[client/src/main/java/com/vaadin/client/connectors/treegrid/TreeGridConnector.java, line 257 at r1](https://reviewable.io:443/reviews/vaadin/framework/8906#-KfpikWM9wnkSYdRN0Jj:-KfpikWM9wnkSYdRN0Jk:bd3bloc) ([raw file](https://github.com/vaadin/framework/blob/26deef47d99c94eca63e98a555f949eb382a48ca/client/src/main/java/com/vaadin/client/connectors/treegrid/TreeGridConnector.java#L257)):*
> ```Java
>             }
> 
>             // Navigate within hierarchy with ALT/OPTION + ARROW KEY when
> ```

comment is no longer accurate

---

*[client/src/main/java/com/vaadin/client/connectors/treegrid/TreeGridConnector.java, line 274 at r1](https://reviewable.io:443/reviews/vaadin/framework/8906#-Kfpiu0zLRdck0cvP-ng:-Kfpiu1-iyVBflkQ6Xnx:b-jdb5ln) ([raw file](https://github.com/vaadin/framework/blob/26deef47d99c94eca63e98a555f949eb382a48ca/client/src/main/java/com/vaadin/client/connectors/treegrid/TreeGridConnector.java#L274)):*
> ```Java
>                     switch (domEvent.getKeyCode()) {
>                     case KeyCodes.KEY_RIGHT:
>                         if (!leaf && collapsedOrLeaf) {
> ```

some of these conditions get hard to read when part of the logical operations are separate and part in the name - I'd still keep these separate

---

*[client/src/main/java/com/vaadin/client/connectors/treegrid/TreeGridConnector.java, line 279 at r1](https://reviewable.io:443/reviews/vaadin/framework/8906#-KfpjCr6QHV64-NXO-ee:-KfpjCr6QHV64-NXO-ef:b3woe28) ([raw file](https://github.com/vaadin/framework/blob/26deef47d99c94eca63e98a555f949eb382a48ca/client/src/main/java/com/vaadin/client/connectors/treegrid/TreeGridConnector.java#L279)):*
> ```Java
>                                     event.getCell().getRowIndex(), false);
>                         } else {
>                             // navigate down
> ```

based on my previous comment: it is not easy for me to see if this else branch is called in the correct situations, but it seems to me that we don't necessarily always want to do this on right arrow

---

*[client/src/main/java/com/vaadin/client/connectors/treegrid/TreeGridConnector.java, line 284 at r1](https://reviewable.io:443/reviews/vaadin/framework/8906#-KfpjcmGhKgntXAuWRDi:-KfpjcmGhKgntXAuWRDj:b-re4rsn) ([raw file](https://github.com/vaadin/framework/blob/26deef47d99c94eca63e98a555f949eb382a48ca/client/src/main/java/com/vaadin/client/connectors/treegrid/TreeGridConnector.java#L284)):*
> ```Java
>                         break;
>                     case KeyCodes.KEY_LEFT:
>                         if (collapsedOrLeaf ) {
> ```

extra whitespace; see my previous comments about conditions and readability

---

*[client/src/main/java/com/vaadin/client/connectors/treegrid/TreeGridConnector.java, line 286 at r1](https://reviewable.io:443/reviews/vaadin/framework/8906#-KfpjnfY8a1OJWcvTPlt:-KfpjnfY8a1OJWcvTPlu:b-fq607k) ([raw file](https://github.com/vaadin/framework/blob/26deef47d99c94eca63e98a555f949eb382a48ca/client/src/main/java/com/vaadin/client/connectors/treegrid/TreeGridConnector.java#L286)):*
> ```Java
>                         if (collapsedOrLeaf ) {
>                             // navigate up
>                             getWidget().focusCell(Math.max(0, event.getCell().getRowIndex() - 1), event.getCell().getColumnIndex());
> ```

my impression was that this should go to the parent if collapsed and collapse if expanded - if there is no parent, no action should be taken

---

*[client/src/main/java/com/vaadin/client/widget/treegrid/TreeGrid.java, line 48 at r1](https://reviewable.io:443/reviews/vaadin/framework/8906#-KfpkZKmIkvQfyvG61gG:-KfpkZKmIkvQfyvG61gH:bv9hv00) ([raw file](https://github.com/vaadin/framework/blob/26deef47d99c94eca63e98a555f949eb382a48ca/client/src/main/java/com/vaadin/client/widget/treegrid/TreeGrid.java#L48)):*
> ```Java
>     @Override
>     protected String getFocusPrimaryStyleName() {
>         return "v-tree-grid";
> ```

hard coded might be problematic when the user wants to change the primary style name - or does it really have to be this?

---

*[uitest/src/test/java/com/vaadin/tests/components/treegrid/TreeGridBasicFeaturesTest.java, line 84 at r1](https://reviewable.io:443/reviews/vaadin/framework/8906#-KfplRsJy0rEXIe-Eq_7:-KfplRsJy0rEXIe-Eq_8:b-8myygj) ([raw file](https://github.com/vaadin/framework/blob/26deef47d99c94eca63e98a555f949eb382a48ca/uitest/src/test/java/com/vaadin/tests/components/treegrid/TreeGridBasicFeaturesTest.java#L84)):*
> ```Java
> 
>     @Test
>     public void keyboard_navigation() {
> ```

(some changes may be needed based on the feedback in TGC)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/8906)*
<!-- Sent from Reviewable.io -->

ahie commented:
A couple things I noticed while testing:

* when pressing left on any root level node, the navigation should jump to the nodes parent
* pressing right shouldn't move the focus down

Additionally, this functionality could be documented in `documentation/components/components-treegrid.asciidoc`.

---

Review status: 1 of 5 files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/8906#-:-KfpmNrdZXyginX7pF3D:bv05ltm)*
<!-- Sent from Reviewable.io -->

Commit:
Implement keyboard navigation from collapsed to parent

Commit:
Styles naming changed

Commit:
Short tree test

elmot commented:
> pressing right shouldn't move the focus down
Actually different software behaves different ways. Now the behavior is changed to be exactly as in ticket, and there should be a way to customize that.

---

Review status: 0 of 10 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[client/src/main/java/com/vaadin/client/connectors/treegrid/TreeGridConnector.java, line 274 at r1](https://reviewable.io:443/reviews/vaadin/framework/8906#-Kfpiu0zLRdck0cvP-ng:-KfzcJQfswbPs6uUb4tj:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/26deef47d99c94eca63e98a555f949eb382a48ca/client/src/main/java/com/vaadin/client/connectors/treegrid/TreeGridConnector.java#L274)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

some of these conditions get hard to read when part of the logical operations are separate and part in the name - I'd still keep these separate
</blockquote></details>

Done.

---

*[client/src/main/java/com/vaadin/client/connectors/treegrid/TreeGridConnector.java, line 279 at r1](https://reviewable.io:443/reviews/vaadin/framework/8906#-KfpjCr6QHV64-NXO-ee:-KfzcMqNjqzZEyHpfvy4:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/26deef47d99c94eca63e98a555f949eb382a48ca/client/src/main/java/com/vaadin/client/connectors/treegrid/TreeGridConnector.java#L279)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

based on my previous comment: it is not easy for me to see if this else branch is called in the correct situations, but it seems to me that we don't necessarily always want to do this on right arrow
</blockquote></details>

Done.

---

*[client/src/main/java/com/vaadin/client/connectors/treegrid/TreeGridConnector.java, line 284 at r1](https://reviewable.io:443/reviews/vaadin/framework/8906#-KfpjcmGhKgntXAuWRDi:-KfzcPECeUgDNVc9mhh2:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/26deef47d99c94eca63e98a555f949eb382a48ca/client/src/main/java/com/vaadin/client/connectors/treegrid/TreeGridConnector.java#L284)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

extra whitespace; see my previous comments about conditions and readability
</blockquote></details>

Done.

---

*[client/src/main/java/com/vaadin/client/connectors/treegrid/TreeGridConnector.java, line 286 at r1](https://reviewable.io:443/reviews/vaadin/framework/8906#-KfpjnfY8a1OJWcvTPlt:-KfzcPyfV2NtatRnkupt:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/26deef47d99c94eca63e98a555f949eb382a48ca/client/src/main/java/com/vaadin/client/connectors/treegrid/TreeGridConnector.java#L286)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

my impression was that this should go to the parent if collapsed and collapse if expanded - if there is no parent, no action should be taken
</blockquote></details>

Done.

---

*[client/src/main/java/com/vaadin/client/widget/treegrid/TreeGrid.java, line 48 at r1](https://reviewable.io:443/reviews/vaadin/framework/8906#-KfpkZKmIkvQfyvG61gG:-KfzcRB0vO83pbailCW7:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/26deef47d99c94eca63e98a555f949eb382a48ca/client/src/main/java/com/vaadin/client/widget/treegrid/TreeGrid.java#L48)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

hard coded might be problematic when the user wants to change the primary style name - or does it really have to be this?
</blockquote></details>

Done.

---

*[uitest/src/test/java/com/vaadin/tests/components/treegrid/TreeGridBasicFeaturesTest.java, line 84 at r1](https://reviewable.io:443/reviews/vaadin/framework/8906#-KfplRsJy0rEXIe-Eq_7:-KfzdE5WD6P5enEZYIVk:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/26deef47d99c94eca63e98a555f949eb382a48ca/uitest/src/test/java/com/vaadin/tests/components/treegrid/TreeGridBasicFeaturesTest.java#L84)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

(some changes may be needed based on the feedback in TGC)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/8906#-:-Kfq47bdrQzDMOfN5hk0:bz67xxm)*
<!-- Sent from Reviewable.io -->

elmot commented:
Review status: 0 of 10 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[a discussion](https://reviewable.io:443/reviews/vaadin/framework/8906#-KfzdIcdVjbIBfwwswf4:-KfzdIcdVjbIBfwwswf5:b-in7uy2) (no related file):*
This change needs a big tree list. Please do not merge until it's done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/8906)*
<!-- Sent from Reviewable.io -->

hesara commented:
Reviewed 3 of 10 files at r2.
Review status: 3 of 10 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[client/src/main/java/com/vaadin/client/connectors/treegrid/TreeGridConnector.java, line 277 at r2](https://reviewable.io:443/reviews/vaadin/framework/8906#-KfzeDCEz1I_vpM8wIAn:-KfzeDCEz1I_vpM8wIAo:b-t90lzx) ([raw file](https://github.com/vaadin/framework/blob/b7a861bc26e9a66d747282765332ad6c94451dbf/client/src/main/java/com/vaadin/client/connectors/treegrid/TreeGridConnector.java#L277)):*
> ```Java
>                     boolean leaf = rowDescription.getBoolean(
>                             TreeGridCommunicationConstants.ROW_LEAF);
>                     if (!leaf) {
> ```

Does navigation left from a leaf to its parent work now?

---

*[client/src/main/java/com/vaadin/client/connectors/treegrid/TreeGridConnector.java, line 281 at r2](https://reviewable.io:443/reviews/vaadin/framework/8906#-KfzeR31JOT9tDtea-IZ:-KfzeR31JOT9tDtea-I_:bvtnja2) ([raw file](https://github.com/vaadin/framework/blob/b7a861bc26e9a66d747282765332ad6c94451dbf/client/src/main/java/com/vaadin/client/connectors/treegrid/TreeGridConnector.java#L281)):*
> ```Java
>                         switch (domEvent.getKeyCode()) {
>                             case KeyCodes.KEY_RIGHT:
>                                 if (collapsed) {
> ```

Do we navigate to the first child if not collapsed or not? Maybe the previous else could just be converted to else if not (leaf) if that condition is not above.

---

*[server/src/main/java/com/vaadin/data/provider/HierarchicalDataCommunicator.java, line 394 at r2](https://reviewable.io:443/reviews/vaadin/framework/8906#-KfzcM8l99iEOXe2oBXd:-KfzcM8l99iEOXe2oBXe:bejmywg) ([raw file](https://github.com/vaadin/framework/blob/b7a861bc26e9a66d747282765332ad6c94451dbf/server/src/main/java/com/vaadin/data/provider/HierarchicalDataCommunicator.java#L394)):*
> ```Java
>      * @param rowIndex the row index
>      * @return the parent index or {@code null} for top-level items
>      * @since
> ```

whole HDC has @since 8.1 so this since tag is not needed

---

*[server/src/main/java/com/vaadin/data/provider/HierarchyMapper.java, line 450 at r2](https://reviewable.io:443/reviews/vaadin/framework/8906#-KfzcWMBcEyCKlEyk-i6:-KfzcWMBcEyCKlEyk-i7:b-yqy8in) ([raw file](https://github.com/vaadin/framework/blob/b7a861bc26e9a66d747282765332ad6c94451dbf/server/src/main/java/com/vaadin/data/provider/HierarchyMapper.java#L450)):*
> ```Java
>      * @param rowIndex the row index
>      * @return the parent index or {@code null} for top-level items
>      * @since
> ```

whole HierarchyMapper has @since 8.1 so this since tag is not needed

---

*[server/src/main/java/com/vaadin/ui/TreeGrid.java, line 201 at r2](https://reviewable.io:443/reviews/vaadin/framework/8906#-KfzclN6MSBtPpYBiYVQ:-KfzclN6MSBtPpYBiYVR:bnuclh) ([raw file](https://github.com/vaadin/framework/blob/b7a861bc26e9a66d747282765332ad6c94451dbf/server/src/main/java/com/vaadin/ui/TreeGrid.java#L201)):*
> ```Java
>                 Integer parentIndex = getDataCommunicator().getParentIndex(rowIndex);
>                 if (parentIndex != null)
>                     getRpcProxy(FocusRpc.class).focusRow(
> ```

(row splitting looks different from what I'd expect from Eclipse)

---

*[shared/src/main/java/com/vaadin/shared/ui/treegrid/FocusParentRpc.java, line 24 at r2](https://reviewable.io:443/reviews/vaadin/framework/8906#-KfzcskZ66zuexVmI5uc:-KfzcskZ66zuexVmI5ud:b-reyfww) ([raw file](https://github.com/vaadin/framework/blob/b7a861bc26e9a66d747282765332ad6c94451dbf/shared/src/main/java/com/vaadin/shared/ui/treegrid/FocusParentRpc.java#L24)):*
> ```Java
>  *
>  * @author Vaadin Ltd
>  * @since
> ```

can already fill in 8.1here

---

*[shared/src/main/java/com/vaadin/shared/ui/treegrid/FocusRpc.java, line 24 at r2](https://reviewable.io:443/reviews/vaadin/framework/8906#-KfzcyXP2W4w2F_BCXmR:-KfzcyXP2W4w2F_BCXmS:b1e5fdc) ([raw file](https://github.com/vaadin/framework/blob/b7a861bc26e9a66d747282765332ad6c94451dbf/shared/src/main/java/com/vaadin/shared/ui/treegrid/FocusRpc.java#L24)):*
> ```Java
>  *
>  * @author Vaadin Ltd
>  * @since
> ```

can already fill in 8.1 here

---

*[shared/src/main/java/com/vaadin/shared/ui/treegrid/FocusRpc.java, line 33 at r2](https://reviewable.io:443/reviews/vaadin/framework/8906#-Kfzd5-ZqG8JUPCvfWPL:-Kfzd5-ZqG8JUPCvfWPM:b-euid7e) ([raw file](https://github.com/vaadin/framework/blob/b7a861bc26e9a66d747282765332ad6c94451dbf/shared/src/main/java/com/vaadin/shared/ui/treegrid/FocusRpc.java#L33)):*
> ```Java
>      * @param rowIndex
>      *            the row index
>      * @param cellIndex
> ```

If this is TreeGrid specific rather than for all Grids, do we really need cellIndex? On the other hand, if we want to generalize this later, we might want to keep it.

---

*[uitest/src/test/java/com/vaadin/tests/components/treegrid/TreeGridBasicFeaturesTest.java, line 43 at r2](https://reviewable.io:443/reviews/vaadin/framework/8906#-KfzdQ4GbzzaewFpkrzd:-KfzdQ4HiNfAe9Kg5-1z:b-z061ma) ([raw file](https://github.com/vaadin/framework/blob/b7a861bc26e9a66d747282765332ad6c94451dbf/uitest/src/test/java/com/vaadin/tests/components/treegrid/TreeGridBasicFeaturesTest.java#L43)):*
> ```Java
>     @Before
>     public void before() {
>         openTestURL("theme=valo","debug");
> ```

(could also just call setDebug(true) before openTestUrl() to do the same)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/8906)*
<!-- Sent from Reviewable.io -->

Commit:
Documentation update

Commit:
`since` fixes, formatting, minor change

elmot commented:
Review status: 3 of 10 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[client/src/main/java/com/vaadin/client/connectors/treegrid/TreeGridConnector.java, line 277 at r2](https://reviewable.io:443/reviews/vaadin/framework/8906#-KfzeDCEz1I_vpM8wIAn:-Kfzi-eDth3eUjO3fpTk:btylw0a) ([raw file](https://github.com/vaadin/framework/blob/b7a861bc26e9a66d747282765332ad6c94451dbf/client/src/main/java/com/vaadin/client/connectors/treegrid/TreeGridConnector.java#L277)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

Does navigation left from a leaf to its parent work now?
</blockquote></details>

Not, and the ticket does not implies that

---

*[client/src/main/java/com/vaadin/client/connectors/treegrid/TreeGridConnector.java, line 281 at r2](https://reviewable.io:443/reviews/vaadin/framework/8906#-KfzeR31JOT9tDtea-IZ:-KfziL4bsMha0sBS4d5X:btylw0a) ([raw file](https://github.com/vaadin/framework/blob/b7a861bc26e9a66d747282765332ad6c94451dbf/client/src/main/java/com/vaadin/client/connectors/treegrid/TreeGridConnector.java#L281)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

Do we navigate to the first child if not collapsed or not? Maybe the previous else could just be converted to else if not (leaf) if that condition is not above.
</blockquote></details>

Not, and the ticket does not implies that

---

*[server/src/main/java/com/vaadin/data/provider/HierarchicalDataCommunicator.java, line 394 at r2](https://reviewable.io:443/reviews/vaadin/framework/8906#-KfzcM8l99iEOXe2oBXd:-KfziVpdR7LsEeMCp6dR:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/b7a861bc26e9a66d747282765332ad6c94451dbf/server/src/main/java/com/vaadin/data/provider/HierarchicalDataCommunicator.java#L394)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

whole HDC has @since 8.1 so this since tag is not needed
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/data/provider/HierarchyMapper.java, line 450 at r2](https://reviewable.io:443/reviews/vaadin/framework/8906#-KfzcWMBcEyCKlEyk-i6:-Kfzia1AgEhRyScDvggg:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/b7a861bc26e9a66d747282765332ad6c94451dbf/server/src/main/java/com/vaadin/data/provider/HierarchyMapper.java#L450)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

whole HierarchyMapper has @since 8.1 so this since tag is not needed
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/ui/TreeGrid.java, line 201 at r2](https://reviewable.io:443/reviews/vaadin/framework/8906#-KfzclN6MSBtPpYBiYVQ:-Kfziq3g6_XlhVf0ZU9J:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/b7a861bc26e9a66d747282765332ad6c94451dbf/server/src/main/java/com/vaadin/ui/TreeGrid.java#L201)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

(row splitting looks different from what I'd expect from Eclipse)
</blockquote></details>

Done.

---

*[shared/src/main/java/com/vaadin/shared/ui/treegrid/FocusParentRpc.java, line 24 at r2](https://reviewable.io:443/reviews/vaadin/framework/8906#-KfzcskZ66zuexVmI5uc:-Kfziz8FAAW7dQxJrwot:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/b7a861bc26e9a66d747282765332ad6c94451dbf/shared/src/main/java/com/vaadin/shared/ui/treegrid/FocusParentRpc.java#L24)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

can already fill in 8.1here
</blockquote></details>

Done.

---

*[shared/src/main/java/com/vaadin/shared/ui/treegrid/FocusRpc.java, line 24 at r2](https://reviewable.io:443/reviews/vaadin/framework/8906#-KfzcyXP2W4w2F_BCXmR:-KfzizazgwMALVhVrU_2:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/b7a861bc26e9a66d747282765332ad6c94451dbf/shared/src/main/java/com/vaadin/shared/ui/treegrid/FocusRpc.java#L24)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

can already fill in 8.1 here
</blockquote></details>

Done.

---

*[shared/src/main/java/com/vaadin/shared/ui/treegrid/FocusRpc.java, line 33 at r2](https://reviewable.io:443/reviews/vaadin/framework/8906#-Kfzd5-ZqG8JUPCvfWPL:-Kfzj0CGPa3kzK7rCCXD:bfme5hr) ([raw file](https://github.com/vaadin/framework/blob/b7a861bc26e9a66d747282765332ad6c94451dbf/shared/src/main/java/com/vaadin/shared/ui/treegrid/FocusRpc.java#L33)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

If this is TreeGrid specific rather than for all Grids, do we really need cellIndex? On the other hand, if we want to generalize this later, we might want to keep it.
</blockquote></details>

Yes, exactly, I think we don't have to close that door.

---

*[uitest/src/test/java/com/vaadin/tests/components/treegrid/TreeGridBasicFeaturesTest.java, line 43 at r2](https://reviewable.io:443/reviews/vaadin/framework/8906#-KfzdQ4GbzzaewFpkrzd:-KfzjNbeI3jinlwQ97M6:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/b7a861bc26e9a66d747282765332ad6c94451dbf/uitest/src/test/java/com/vaadin/tests/components/treegrid/TreeGridBasicFeaturesTest.java#L43)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

(could also just call setDebug(true) before openTestUrl() to do the same)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/8906)*
<!-- Sent from Reviewable.io -->

hesara commented:
Reviewed 6 of 7 files at r3.
Review status: 8 of 11 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/8906)*
<!-- Sent from Reviewable.io -->

elmot commented:
Review status: 8 of 11 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[a discussion](https://reviewable.io:443/reviews/vaadin/framework/8906#-KfzdIcdVjbIBfwwswf4:-Kg-N6hO8TkiuIyViPQe:b-rxgi1n) (no related file):*
<details><summary><i>Previously, elmot (Ilia Motornyi) wrote…</i></summary><blockquote>

This change needs a big tree list. Please do not merge until it's done.
</blockquote></details>

There is a bug #8840, the test can not be done because of that.  There will be another PR and ticket.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/8906)*
<!-- Sent from Reviewable.io -->

Commit:
Test is disabled because of #8840

Commit:
Fix after merge

hesara commented:
Reviewed 4 of 7 files at r4.
Review status: 8 of 11 files reviewed at latest revision, 4 unresolved discussions.

---

*[documentation/components/components-treegrid.asciidoc, line 115 at r3](https://reviewable.io:443/reviews/vaadin/framework/8906#-KfzuQtArGzFdt0Vd_zp:-KfzuQtBXS34AHcuG61O:bz4nm1d) ([raw file](https://github.com/vaadin/framework/blob/a9588dfeebbff8edb44ab4a92afaca07c460748e/documentation/components/components-treegrid.asciidoc#L115)):*
> ```AsciiDoc
> 
> As opposed to [classname]#Grid#, individual cell are not focusable in [classname]#TreeGrid#, and only whole rows
> receive focus. User can navigate through rows with kbd:[Up] and kbd:[Down], collapse rows with kbd:[Left],
> ```

(I'd say "The user")

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/8906)*
<!-- Sent from Reviewable.io -->

ahie commented:
Review status: 8 of 11 files reviewed at latest revision, 5 unresolved discussions.

---

*[documentation/components/components-treegrid.asciidoc, line 114 at r3](https://reviewable.io:443/reviews/vaadin/framework/8906#-Kg-Yon4HamMM5mZxrOp:-Kg-Yon4HamMM5mZxrOq:b-ctmct4) ([raw file](https://github.com/vaadin/framework/blob/a9588dfeebbff8edb44ab4a92afaca07c460748e/documentation/components/components-treegrid.asciidoc#L114)):*
> cell

cells

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/8906)*
<!-- Sent from Reviewable.io -->

Commit:
Dosc fixed

elmot commented:
Review status: 8 of 11 files reviewed at latest revision, 5 unresolved discussions.

---

*[documentation/components/components-treegrid.asciidoc, line 114 at r3](https://reviewable.io:443/reviews/vaadin/framework/8906#-Kg-Yon4HamMM5mZxrOp:-Kg-fVEaUC_k47YXybQg:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/a9588dfeebbff8edb44ab4a92afaca07c460748e/documentation/components/components-treegrid.asciidoc#L114)):*
<details><summary><i>Previously, ahie (Aleksi Hietanen) wrote…</i></summary><blockquote>

> cell

cells
</blockquote></details>

Done.

---

*[documentation/components/components-treegrid.asciidoc, line 115 at r3](https://reviewable.io:443/reviews/vaadin/framework/8906#-KfzuQtArGzFdt0Vd_zp:-Kg-fVY5npCkbI1hYyh2:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/a9588dfeebbff8edb44ab4a92afaca07c460748e/documentation/components/components-treegrid.asciidoc#L115)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

(I'd say "The user")
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/8906)*
<!-- Sent from Reviewable.io -->

Commit:
Re-enable test for tree grid navigation

hesara commented:
Reviewed 1 of 1 files at r5.
Review status: 9 of 11 files reviewed at latest revision, 2 unresolved discussions.

---

*[client/src/main/java/com/vaadin/client/connectors/treegrid/TreeGridConnector.java, line 277 at r2](https://reviewable.io:443/reviews/vaadin/framework/8906#-KfzeDCEz1I_vpM8wIAn:-KgEa9ChxC6iNQZ-ULO0:b1n69eb) ([raw file](https://github.com/vaadin/framework/blob/b7a861bc26e9a66d747282765332ad6c94451dbf/client/src/main/java/com/vaadin/client/connectors/treegrid/TreeGridConnector.java#L277)):*
<details><summary><i>Previously, elmot (Ilia Motornyi) wrote…</i></summary><blockquote>

Not, and the ticket does not implies that
</blockquote></details>

IMO this must be addressed - looking into this

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/8906)*
<!-- Sent from Reviewable.io -->

ahie committed:
Fix focusing on parent, navigation from leaf nodes

ahie committed:
Merge remote-tracking branch 'github/8939_huge-treegrid-keyboard-navigation' into 8758_TreeGrid_keyb_navigation

ahie committed:
Merge remote-tracking branch 'github/master' into 8758_TreeGrid_keyb_navigation

hesara commented:
Reviewed 2 of 2 files at r6, 2 of 3 files at r7.
Review status: 11 of 13 files reviewed at latest revision, 4 unresolved discussions.

---

*[themes/src/main/themes/VAADIN/themes/valo/components/_treegrid.scss, line 16 at r6](https://reviewable.io:443/reviews/vaadin/framework/8906#-KgJd4TDIQcViRXqxxiA:-KgJd4TDIQcViRXqxxiB:bnii9r7) ([raw file](https://github.com/vaadin/framework/blob/252a67a258556b5d60441d7ff2bc84c2c9a11783/themes/src/main/themes/VAADIN/themes/valo/components/_treegrid.scss#L16)):*
> ```SCSS
> 
> 
> @mixin treegrid {
> ```

(this really should be parameterized with the primary style name, but that is covered by #8930)

---

*[themes/src/main/themes/VAADIN/themes/valo/components/_treegrid.scss, line 69 at r6](https://reviewable.io:443/reviews/vaadin/framework/8906#-KgJdOOe22KhgOOQZOHf:-KgJdOOe22KhgOOQZOHg:bm6e6gj) ([raw file](https://github.com/vaadin/framework/blob/252a67a258556b5d60441d7ff2bc84c2c9a11783/themes/src/main/themes/VAADIN/themes/valo/components/_treegrid.scss#L69)):*
> ```SCSS
>   }
> 
>   .v-grid-rowmode-row-focused {
> ```

(as above - these probably should be based on the primary style name, and the grid mixin should be included with it v-tree-grid as primary style name)

---

*[uitest/src/test/java/com/vaadin/tests/components/treegrid/TreeGridHugeTreeNavigationTest.java, line 90 at r7](https://reviewable.io:443/reviews/vaadin/framework/8906#-KgJcumVmtd_IFWbfTyz:-KgJcumWfv5mGloc94zI:b-tzxbwe) ([raw file](https://github.com/vaadin/framework/blob/0be7937abdd691e35bcb1014d8f956c2e49e364e/uitest/src/test/java/com/vaadin/tests/components/treegrid/TreeGridHugeTreeNavigationTest.java#L90)):*
> ```Java
>     }
> 
>     private WebElement gindFocusedRow() {
> ```

findFocusedRow()

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/8906)*
<!-- Sent from Reviewable.io -->

ahie commented:
Review status: 10 of 13 files reviewed at latest revision, 2 unresolved discussions.

---

*[uitest/src/test/java/com/vaadin/tests/components/treegrid/TreeGridHugeTreeNavigationTest.java, line 90 at r7](https://reviewable.io:443/reviews/vaadin/framework/8906#-KgJcumVmtd_IFWbfTyz:-KgJeQWBEB2xKdninaGL:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/0be7937abdd691e35bcb1014d8f956c2e49e364e/uitest/src/test/java/com/vaadin/tests/components/treegrid/TreeGridHugeTreeNavigationTest.java#L90)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

findFocusedRow()
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/8906)*
<!-- Sent from Reviewable.io -->

hesara commented:
Reviewed 1 of 3 files at r7, 1 of 1 files at r8.
Review status: 12 of 13 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/8906)*
<!-- Sent from Reviewable.io -->

ahie committed:
Minor improvements to TreeGridHugeTreeNavigationTest

ahie committed:
Add FunctionalInterface annotation to FocusRpc

ahie committed:
Fix Escalator#paintRemoveRows when new row count drops below current escalator rows

Also fixes other bugs discovered in TreeGrid keyboard navigation. Tests
in TreeGridHugeTreeNavigationTest cover the fixed cases.

Commit:
If focused row isn't present in cache, scroll it into view + some renames.

Commit:
Forbidden collapse support added

elmot commented:
Review status: 6 of 15 files reviewed at latest revision, 1 unresolved discussion.

---

*[client/src/main/java/com/vaadin/client/connectors/treegrid/TreeGridConnector.java, line 277 at r2](https://reviewable.io:443/reviews/vaadin/framework/8906#-KfzeDCEz1I_vpM8wIAn:-KgtPXD7CCLBES-UxJlD:bb6un1j) ([raw file](https://github.com/vaadin/framework/blob/b7a861bc26e9a66d747282765332ad6c94451dbf/client/src/main/java/com/vaadin/client/connectors/treegrid/TreeGridConnector.java#L277)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

IMO this must be addressed - looking into this
</blockquote></details>

Done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/8906)*
<!-- Sent from Reviewable.io -->

ahie commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, someone else should review the Escalator changes.

---

Reviewed 1 of 7 files at r4, 1 of 6 files at r10.
Review status: 8 of 15 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/8906#-:-KgwmwMuB2cXjs4y1kus:b-xwuxxp)*
<!-- Sent from Reviewable.io -->

ahie commented:
Reviewed 1 of 6 files at r10.
Review status: 9 of 15 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/8906)*
<!-- Sent from Reviewable.io -->

tsuoanttila commented:
Reviewed 1 of 5 files at r9.
Review status: 10 of 15 files reviewed at latest revision, 2 unresolved discussions.

---

*[client/src/main/java/com/vaadin/client/widgets/Escalator.java, line 3123 at r10](https://reviewable.io:443/reviews/vaadin/framework/8906#-Kgx4oLnDigTQ4GSL3el:-Kgx4oLnDigTQ4GSL3em:b7ka2r8) ([raw file](https://github.com/vaadin/framework/blob/eb4716f048c01a65ce87486df239e40dd53da6a0/client/src/main/java/com/vaadin/client/widgets/Escalator.java#L3123)):*
> ```Java
>                     }
> 
>                     moveAndUpdateEscalatorRows(
> ```

What are the parameters being passed to this method? what is the use of each of those? this code is practically no longer documented by comments and has become a bit trickier to understand without looking into the moveAndUpdateEscalatorRows method.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/8906)*
<!-- Sent from Reviewable.io -->

tsuoanttila commented:
Reviewed 1 of 5 files at r9, 4 of 6 files at r10.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[server/src/main/java/com/vaadin/data/provider/HierarchicalDataCommunicator.java, line 360 at r9](https://reviewable.io:443/reviews/vaadin/framework/8906#-Kgx5wldu3-O-_bHUPR3:-Kgx5wldu3-O-_bHUPR4:b-cioao8) ([raw file](https://github.com/vaadin/framework/blob/12799349b4e8c5872f4e9cf2add718f9779d6e72/server/src/main/java/com/vaadin/data/provider/HierarchicalDataCommunicator.java#L360)):*
> ```Java
>         if (mapper.isCollapsed(collapsedRowKey)) {
>             return false;
>         } else {
> ```

see doExpand comment.

---

*[server/src/main/java/com/vaadin/data/provider/HierarchicalDataCommunicator.java, line 402 at r9](https://reviewable.io:443/reviews/vaadin/framework/8906#-Kgx5qMLd6js66dcA-Lb:-Kgx5qMLd6js66dcA-Lc:budkedt) ([raw file](https://github.com/vaadin/framework/blob/12799349b4e8c5872f4e9cf2add718f9779d6e72/server/src/main/java/com/vaadin/data/provider/HierarchicalDataCommunicator.java#L402)):*
>  else {
unnecessary else block and indent. if block returns -> no need for else

---

*[uitest/src/test/java/com/vaadin/tests/components/treegrid/TreeGridHugeTreeNavigationTest.java, line 19 at r10](https://reviewable.io:443/reviews/vaadin/framework/8906#-Kgx6J_YQWq8aJgChxBz:-Kgx6J_YQWq8aJgChxC-:b-3tno7k) ([raw file](https://github.com/vaadin/framework/blob/eb4716f048c01a65ce87486df239e40dd53da6a0/uitest/src/test/java/com/vaadin/tests/components/treegrid/TreeGridHugeTreeNavigationTest.java#L19)):*
> ```Java
> import com.vaadin.tests.tb3.ParameterizedTB3Runner;
> 
> @RunWith(ParameterizedTB3Runner.class)
> ```

With a quick look I did not see any parameters on this test. Did I just miss it, or is this unnecessary?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/8906)*
<!-- Sent from Reviewable.io -->

Commit:
Fixes after a review

Commit:
Test code simplified

tsuoanttila commented:
Looks ok for me

---

Reviewed 3 of 3 files at r11.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/8906#-:-KgxNr_Qy-sM1jyte3N1:b48izfm)*
<!-- Sent from Reviewable.io -->

elmot commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[client/src/main/java/com/vaadin/client/widgets/Escalator.java, line 3123 at r10](https://reviewable.io:443/reviews/vaadin/framework/8906#-Kgx4oLnDigTQ4GSL3el:-KgxLmpK2IpBfUhS_Uwq:b-47fglo) ([raw file](https://github.com/vaadin/framework/blob/eb4716f048c01a65ce87486df239e40dd53da6a0/client/src/main/java/com/vaadin/client/widgets/Escalator.java#L3123)):*
<details><summary><i>Previously, tsuoanttila (Teemu Suo-Anttila) wrote…</i></summary><blockquote>

What are the parameters being passed to this method? what is the use of each of those? this code is practically no longer documented by comments and has become a bit trickier to understand without looking into the moveAndUpdateEscalatorRows method.
</blockquote></details>

Corrected a bit. IDNK if it's enough.

---

*[server/src/main/java/com/vaadin/data/provider/HierarchicalDataCommunicator.java, line 360 at r9](https://reviewable.io:443/reviews/vaadin/framework/8906#-Kgx5wldu3-O-_bHUPR3:-KgxEWrhRL-h9-J_evoL:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/12799349b4e8c5872f4e9cf2add718f9779d6e72/server/src/main/java/com/vaadin/data/provider/HierarchicalDataCommunicator.java#L360)):*
<details><summary><i>Previously, tsuoanttila (Teemu Suo-Anttila) wrote…</i></summary><blockquote>

see doExpand comment.
</blockquote></details>

Done.

---

*[uitest/src/test/java/com/vaadin/tests/components/treegrid/TreeGridHugeTreeNavigationTest.java, line 19 at r10](https://reviewable.io:443/reviews/vaadin/framework/8906#-Kgx6J_YQWq8aJgChxBz:-KgxLigs2lsJa5-UwMhz:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/eb4716f048c01a65ce87486df239e40dd53da6a0/uitest/src/test/java/com/vaadin/tests/components/treegrid/TreeGridHugeTreeNavigationTest.java#L19)):*
<details><summary><i>Previously, tsuoanttila (Teemu Suo-Anttila) wrote…</i></summary><blockquote>

With a quick look I did not see any parameters on this test. Did I just miss it, or is this unnecessary?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/8906)*
<!-- Sent from Reviewable.io -->

elmot commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[server/src/main/java/com/vaadin/data/provider/HierarchicalDataCommunicator.java, line 402 at r9](https://reviewable.io:443/reviews/vaadin/framework/8906#-Kgx5qMLd6js66dcA-Lb:-KgxdPIK6qStKZr-GWC_:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/12799349b4e8c5872f4e9cf2add718f9779d6e72/server/src/main/java/com/vaadin/data/provider/HierarchicalDataCommunicator.java#L402)):*
<details><summary><i>Previously, tsuoanttila (Teemu Suo-Anttila) wrote…</i></summary><blockquote>

>  else {
unnecessary else block and indent. if block returns -> no need for else
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/8906)*
<!-- Sent from Reviewable.io -->
